### PR TITLE
[NativeAOT-LLVM] Merge to Sep 24

### PR DESCRIPTION
**Pull request by yowl**
This PR merges from upstream to Sep 2024.

I made `acdHelper` `public` so we can access it from llvm codegen

See the comments at the end re non-running tests and failure on linux to find `libz`

### CODE REVIEW

**Comment by SingleAccretion** [comment]
> remove ThrowMisalign

This helper is NAOT-LLVM specific and should be kept (it is used for atomics).

**Comment by yowl** [comment]
Thanks, yes realised that a bit later.

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by yowl** [comment]
Compiling XUnitLogChecker is failing on `linux-x64`, first on `bfd`/`ld` decision, fixing that moves it on to missing `-lz` library.  Wouldn't have expected this from a merge.  Perhaps the container is not correct for some reason.

**Comment by yowl** [comment]
cc @dotnet/nativeaot-llvm 

There's a few strange things here:

1. The tests for browser and wasi claim to pass, but in fact they are not running.  I think the merging `nativeaot` wrapper should be compiled for the host so it can write files, but maybe I misunderstood the discussion in https://github.com/dotnet/runtimelab/issues/2695
2. I've had to remove the use of `-lz` on linux for the building of tests.  I can't see why this works upstream, but not here.

File: src/coreclr/jit/utils.cpp [hunk 1]
```diff
@@ -1626,24 +1626,13 @@ void HelperCallProperties::init()
...
                 isPure        = true;
                 noThrow       = true;
                 nonNullReturn = true;
                 break;
 
-            case CORINFO_HELP_READYTORUN_GENERIC_HANDLE:
-                // Typically this is a no-throw call, but may become always-throw when fetching an unavailable handle.
-                isPure        = true;
-                nonNullReturn = true;
-                break;
```
Comment (SingleAccretion) [inline src/coreclr/jit/utils.cpp]: This should be restored. `CORINFO_HELP_READYTORUN_GENERIC_HANDLE` may throw. This should fail one of the tests (`UnitTests` if memory serves me right).

File: src/libraries/System.Net.Http/src/System.Net.Http.csproj [hunk 2]
```diff
@@ -457,20 +457,15 @@
              Link="Common\System\Net\Http\HttpHandlerDefaults.cs" />
   </ItemGroup>
 
-<<<<<<< HEAD
   <!-- TODO-LLVM: This is not upstreamable because it makes the build runtime-specific. -->
   <ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'wasi' and '$(RuntimeFlavor)' != 'CoreCLR'">
-    <Reference Include="System.Threading.Thread" />
-=======
-  <ItemGroup Condition="'$(TargetPlatformIdentifier)' == 'wasi'">
```
Comment (SingleAccretion) [inline src/libraries/System.Net.Http/src/System.Net.Http.csproj]: Can we delete these ` '$(RuntimeFlavor)' != 'CoreCLR'` special cases now and take upstream verbatim? It would really like to avoid NAOT-LLVM conditionals outside of core runtime.

File: src/tests/nativeaot/SmokeTests/SharedLibrary/SharedLibrary.csproj [hunk 3]
```diff
@@ -5,26 +5,18 @@
...
-    <CLRTestTargetUnsupported Condition="'$(TargetsAppleMobile)' == 'true'">true</CLRTestTargetUnsupported>
-
-    <!-- https://github.com/dotnet/runtimelab/issues/2648 -->
-    <CLRTestTargetUnsupported Condition="'$(TargetOS)' == 'wasi'">true</CLRTestTargetUnsupported>
-=======
     <BuildAsStandalone>false</BuildAsStandalone>
-    <!-- Unable to compile a project with the Library output type for apple mobile devices -->
-    <CLRTestTargetUnsupported Condition="'$(TargetsAppleMobile)' == 'true'">true</CLRTestTargetUnsupported>
+    <!-- Unable to compile a project with the Library output type for apple mobile devices or WASI -->
+    <CLRTestTargetUnsupported Condition="'$(TargetsAppleMobile)' == 'true' or '$(TargetOS)' == 'wasi'">true</CLRTestTargetUnsupported>
```
Comment (SingleAccretion) [inline src/tests/nativeaot/SmokeTests/SharedLibrary/SharedLibrary.csproj]: This should stay as a distinct condition pointing to `https://github.com/dotnet/runtimelab/issues/2648`.

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/SocketsHttpHandler.cs [hunk 4]
```diff
@@ -213,5 +213,26 @@ public Func<SocketsHttpPlaintextStreamFilterContext, CancellationToken, ValueTas
             get => throw new PlatformNotSupportedException();
             set => throw new PlatformNotSupportedException();
         }
+
+        public static bool SupportsAutomaticDecompression
```
Comment (SingleAccretion) [inline src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/SocketsHttpHandler.cs]: Why do we have these differences from upstream?

File: src/tests/nativeaot/StartupHook/StartupHook.csproj [hunk 5]
```diff
@@ -18,5 +18,6 @@
 
   <PropertyGroup>
     <CLRTestTargetUnsupported Condition="'$(TargetsWasm)' == 'true'">true</CLRTestTargetUnsupported>
+    <CLRTestExitCode>100</CLRTestExitCode>
```
Comment (SingleAccretion) [inline src/tests/nativeaot/StartupHook/StartupHook.csproj]: Why is this needed?

File: src/tests/nativeaot/nativeaot.csproj [hunk 6]
```diff
@@ -1,8 +1,13 @@
 <Project Sdk="Microsoft.NET.Sdk">
-    <ItemGroup>
-      <MergedWrapperProjectReference Include="*/**/*.??proj" Exclude="*/**/Library.csproj" />
-    </ItemGroup>
-  
-    <Import Project="$(TestSourceDir)MergedTestRunner.targets" />
-  </Project>
+  <PropertyGroup>
+    <CLRTestExitCode Condition="'$(CLRTestExitCode)' == ''">100</CLRTestExitCode>
```
Comment (SingleAccretion) [inline src/tests/nativeaot/nativeaot.csproj]: This should already be set by `Common\CLRTest.Execute.targets`, why does it need to be duplicated?

File: src/tests/nativeaot/nativeaot.csproj [hunk 7]
```diff
@@ -1,8 +1,13 @@
 <Project Sdk="Microsoft.NET.Sdk">
-    <ItemGroup>
-      <MergedWrapperProjectReference Include="*/**/*.??proj" Exclude="*/**/Library.csproj" />
-    </ItemGroup>
-  
-    <Import Project="$(TestSourceDir)MergedTestRunner.targets" />
-  </Project>
```
Comment (SingleAccretion) [inline src/tests/nativeaot/nativeaot.csproj]: It would be good to reduce the diff on this file by following the upstream (wrong) whitespace formatting.

File: src/tests/nativeaot/nativeaot.csproj [hunk 8]
```diff
@@ -1,8 +1,13 @@
...
+  <PropertyGroup>
+    <CLRTestExitCode Condition="'$(CLRTestExitCode)' == ''">100</CLRTestExitCode>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <MergedWrapperProjectReference Include="*/**/*.??proj" Exclude="*/**/Library.csproj;*/**/HelloWasm/*.??proj" Condition="'$(TargetsWasm)' != 'true'" />
+    <MergedWrapperProjectReference Include="*/**/*.??proj" Exclude="*/**/Library.csproj"  Condition="'$(TargetsWasm)' == 'true'" />
```
Comment (SingleAccretion) [inline src/tests/nativeaot/nativeaot.csproj]: IMO it would be better to "include everything" (basically upstream include), then exclude the WASM projects we care about, basically replicating the list we had in `src/tests/Common/dirs.proj`.

Also, `dirs.proj` itself needs to be updated to be synced to upstream version now that no changes are needed for it.

File: src/tests/Common/XUnitLogChecker/XUnitLogChecker.csproj [hunk 9]
```diff
@@ -7,7 +7,7 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <PublishAot Condition="'$(NativeAotSupported)' == 'true'">true</PublishAot>
     <SuppressTrimAnalysisWarnings>false</SuppressTrimAnalysisWarnings>
-    <LinkerFlavor Condition="'$(CrossBuild)' == 'true' and '$(TargetsLinux)' == 'true'">lld</LinkerFlavor>
+    <LinkerFlavor Condition="('$(__TestBuildMode)' == 'nativeaot' or '$(CrossBuild)' == 'true') and '$(TargetsLinux)' == 'true'">lld</LinkerFlavor>
```
Comment (SingleAccretion) [inline src/tests/Common/XUnitLogChecker/XUnitLogChecker.csproj]: This indicates our test build isn't passing `CrossBuild` down correctly. I think we are missing `$(crossArg)` in the `./src/tests/build.sh` invocations in the templates. I think this explains the `-lz` mystery issue as well.

**Review by SingleAccretion — COMMENTED** [review]
Obviously, runtime tests need to be fixed to run. I think for this merge we should use the escape hatches, i. e. set `BuildAllTestsAsStandalone` somewhere globally, and make the merged model work separately.

File: src/tests/Common/XUnitLogChecker/XUnitLogChecker.csproj (hunk 9 above)
Comment (yowl) [reply src/tests/Common/XUnitLogChecker/XUnitLogChecker.csproj]: Done.

File: src/tests/nativeaot/nativeaot.csproj (hunk 8 above)
Comment (yowl) [reply src/tests/nativeaot/nativeaot.csproj]: We can't use this for Wasm as we are not using the merged wrapper, but have changed for non -wasm

File: src/tests/nativeaot/nativeaot.csproj (hunk 7 above)
Comment (yowl) [reply src/tests/nativeaot/nativeaot.csproj]: Done

File: src/tests/nativeaot/nativeaot.csproj (hunk 6 above)
Comment (yowl) [reply src/tests/nativeaot/nativeaot.csproj]: removed.

File: src/tests/nativeaot/StartupHook/StartupHook.csproj (hunk 5 above)
Comment (yowl) [reply src/tests/nativeaot/StartupHook/StartupHook.csproj]: reverted

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/SocketsHttpHandler.cs (hunk 4 above)
Comment (yowl) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/SocketsHttpHandler.cs]: wasi doesn't compile without them, let me check why

File: src/tests/nativeaot/StartupHook/StartupHook.csproj (hunk 5 above)
Comment (yowl) [reply src/tests/nativeaot/StartupHook/StartupHook.csproj]: reverted

File: src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/SocketsHttpHandler.cs (hunk 4 above)
Comment (yowl) [reply src/libraries/System.Net.Http/src/System/Net/Http/BrowserHttpHandler/SocketsHttpHandler.cs]: Reverted and remove some workarounds that were pending this merge

File: src/tests/nativeaot/SmokeTests/SharedLibrary/SharedLibrary.csproj (hunk 3 above)
Comment (yowl) [reply src/tests/nativeaot/SmokeTests/SharedLibrary/SharedLibrary.csproj]: THanks, missed that.

File: src/libraries/System.Net.Http/src/System.Net.Http.csproj (hunk 2 above)
Comment (yowl) [reply src/libraries/System.Net.Http/src/System.Net.Http.csproj]: Yes , we can

File: src/coreclr/jit/utils.cpp (hunk 1 above)
Comment (yowl) [reply src/coreclr/jit/utils.cpp]: restored

**Comment by yowl** [comment]
Got there finally.  Can be reviewed again, thanks!

File: src/coreclr/jit/utils.cpp [hunk 10]
```diff
@@ -1629,7 +1629,6 @@ void HelperCallProperties::init()
             case CORINFO_HELP_READYTORUN_GENERIC_HANDLE:
                 // logging helpers are not technically pure but can be optimized away
                 isPure        = true;
-                noThrow       = true;
                 nonNullReturn = true;
                 break;
```
Comment (SingleAccretion) [inline src/coreclr/jit/utils.cpp]:
```suggestion
                noThrow       = true;
                nonNullReturn = true;
                break;

            case CORINFO_HELP_READYTORUN_GENERIC_HANDLE:
                // Typically this is a no-throw call, but may become always-throw when fetching an unavailable handle.
                isPure        = true;
                nonNullReturn = true;
                break;
```
And take out `case CORINFO_HELP_READYTORUN_GENERIC_HANDLE` from above. There is no need to pessimize the other helpers.

File: src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs [hunk 11]
```diff
@@ -167,10 +167,8 @@ protected internal override async Task<HttpResponseMessage> SendAsync(HttpReques
 
         #region PlatformNotSupported
 #pragma warning disable CA1822
-
         internal ClientCertificateOption ClientCertificateOptions;
 
-
```
Comment (SingleAccretion) [inline src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs]: Whitespace changes should be reverted.

File: src/libraries/System.Net.Http/src/System.Net.Http.csproj [hunk 12]
```diff
@@ -24,13 +24,8 @@
...
   </PropertyGroup>
 
-  <!-- TODO-LLVM: This is not upstreamable because it makes the build runtime-specific. -->
-  <PropertyGroup>
-    <DefineConstants Condition="'$(RuntimeFlavor)' == 'CoreCLR'">$(DefineConstants);NATIVE_AOT</DefineConstants>
-  </PropertyGroup>
-  
   <ItemGroup>
-    <ILLinkSubstitutionsXmls Include="$(ILLinkDirectory)ILLink.Substitutions.xml" />
+    <ILLinkSubstitutionsXmls Include="$(ILLinkDirectory)ILLink.Substitutions.xml" Condition="'$(TargetPlatformIdentifier)' != 'wasi'"/>
```
Comment (SingleAccretion) [inline src/libraries/System.Net.Http/src/System.Net.Http.csproj]: `Condition="'$(TargetPlatformIdentifier)' != 'wasi'"` still needed?

File: eng/pipelines/runtimelab/runtimelab-post-build-steps.yml [hunk 13]
```diff
@@ -8,6 +8,7 @@ parameters:
   testFilter: tree nativeaot
   isOfficialBuild: false
   librariesConfiguration: Debug
+  crossArg: $(crossArg)
```
Comment (SingleAccretion) [inline eng/pipelines/runtimelab/runtimelab-post-build-steps.yml]: Why indirect through the `parameters` template arguments, instead of using `$(crossArg)` directly?

File: src/tests/Common/mergedrunnermobile.targets [hunk 14]
```diff
@@ -39,8 +39,26 @@
...
+  
   <!-- Don't use the usual script generation, the scripts it generates are not useful for these tests. -->
-  <Target Name="GenerateExecutionScriptsInternal" />
+  <Target Name="GenerateExecutionScriptsInternal"
+    Condition="$(_CLRTestNeedsToRun) or $(_CLRTestBuildsExecutable)"
+    DependsOnTargets="FetchExternalProperties;Generate@(ExecutionScriptKind, 'ExecutionScript;Generate')ExecutionScript;" />
+
   <Target Name="GenerateLocalRunnerScript" DependsOnTargets="GenerateRunScript" AfterTargets="PublishTestAsSelfContained" />
 
   <Import Project="$(RepositoryEngineeringDir)testing\tests.targets" />
```
Comment (SingleAccretion) [inline src/tests/Common/mergedrunnermobile.targets]: Revert this file?

File: src/tests/Directory.Build.targets [hunk 15]
```diff
@@ -23,6 +23,10 @@
     <IsRuntimeTests>true</IsRuntimeTests>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(TargetsWasm)' == 'true' and '$(RuntimeFlavor)' == 'coreclr'">
+    <BuildAllTestsAsStandalone>true</BuildAllTestsAsStandalone>
+  </PropertyGroup>
```
Comment (SingleAccretion) [inline src/tests/Directory.Build.targets]:
```suggestion
  <PropertyGroup Condition="'$(TargetsWasm)' == 'true' and '$(RuntimeFlavor)' == 'coreclr'">
    <!-- NativeAOT-LLVM: https://github.com/dotnet/runtimelab/issues/2695. -->
    <BuildAllTestsAsStandalone>true</BuildAllTestsAsStandalone>
  </PropertyGroup>  
```

File: src/coreclr/inc/jithelpers.h [hunk 16]
```diff
@@ -119,20 +119,11 @@
...
-    DYNAMICJITHELPER(CORINFO_HELP_OVERFLOW,     NULL,               METHOD__THROWHELPER__THROWOVERFLOWEXCEPTION)
-    DYNAMICJITHELPER(CORINFO_HELP_THROWDIVZERO, NULL,               METHOD__THROWHELPER__THROWDIVIDEBYZEROEXCEPTION)
-    DYNAMICJITHELPER(CORINFO_HELP_THROWNULLREF, NULL,               METHOD__THROWHELPER__THROWNULLREFEXCEPTION)
-    JITHELPER(CORINFO_HELP_THROWMISALIGN,       NULL,               METHOD__NIL)
-    DYNAMICJITHELPER(CORINFO_HELP_VERIFICATION, NULL,               METHOD__THROWHELPER__THROWVERIFICATIONEXCEPTION)
-=======
     DYNAMICJITHELPER(CORINFO_HELP_RNGCHKFAIL,   NULL,               METHOD__THROWHELPERS__THROWINDEXOUTOFRANGEEXCEPTION)
     DYNAMICJITHELPER(CORINFO_HELP_OVERFLOW,     NULL,               METHOD__THROWHELPERS__THROWOVERFLOWEXCEPTION)
     DYNAMICJITHELPER(CORINFO_HELP_THROWDIVZERO, NULL,               METHOD__THROWHELPERS__THROWDIVIDEBYZEROEXCEPTION)
     DYNAMICJITHELPER(CORINFO_HELP_THROWNULLREF, NULL,               METHOD__THROWHELPERS__THROWNULLREFEXCEPTION)
```
Comment (SingleAccretion) [inline src/coreclr/inc/jithelpers.h:125]:
```suggestion
    DYNAMICJITHELPER(CORINFO_HELP_THROWNULLREF, NULL,               METHOD__THROWHELPERS__THROWNULLREFEXCEPTION)
    JITHELPER(CORINFO_HELP_THROWMISALIGN,       NULL,               METHOD__NIL)
```

File: src/tests/nativeaot/nativeaot.csproj [hunk 17]
```diff
@@ -1,8 +1,16 @@
...
+    </ItemGroup>
+
+    <!-- Wasm disabled projects are still in dirs.proj because Wasm is not using this merged test wrapper -->
+
     <ItemGroup>
-      <MergedWrapperProjectReference Include="*/**/*.??proj" Exclude="*/**/Library.csproj" />
+      <MergedWrapperProjectReference Include="*/**/*.??proj" Exclude="@(DisabledProjects)" />
     </ItemGroup>
-  
+
```
Comment (SingleAccretion) [inline src/tests/nativeaot/nativeaot.csproj]:
```suggestion
  
```
Reducing diffs.

File: src/tests/Common/mergedrunnermobile.targets (hunk 14 above)
Comment (yowl) [reply src/tests/Common/mergedrunnermobile.targets]: yes, thanks.

File: src/tests/nativeaot/nativeaot.csproj (hunk 17 above)
Comment (yowl) [reply src/tests/nativeaot/nativeaot.csproj]: thanks, spaes added.

File: eng/pipelines/runtimelab/runtimelab-post-build-steps.yml (hunk 13 above)
Comment (yowl) [reply eng/pipelines/runtimelab/runtimelab-post-build-steps.yml]: no reason, removed.

File: src/libraries/System.Net.Http/src/System.Net.Http.csproj (hunk 12 above)
Comment (yowl) [reply src/libraries/System.Net.Http/src/System.Net.Http.csproj]: no, removed.  Thanks.

File: src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs (hunk 11 above)
Comment (yowl) [reply src/libraries/System.Net.Http/src/System/Net/Http/WasiHttpHandler/WasiHttpHandler.cs]: Done.

File: src/coreclr/jit/utils.cpp (hunk 10 above)
Comment (yowl) [reply src/coreclr/jit/utils.cpp]: Done, thanks.

**Comment by yowl** [comment]
Latest feedback addressed,  thanks!

**Review by SingleAccretion — APPROVED** [review]
LGTM, thank you!

**Comment by jkotas** [comment]
Thank you all!
